apps/src/music/views/SoundsPanel2.tsx

```diff
@@ -0,0 +1,487 @@
+  );
+};
+
+// SoundsPanel: a component used inside a Blockly dropdown that allows
+// for browsing folders and the sounds inside them.
+
+interface SoundsPanelProps {
+  library: MusicLibrary;
+  currentValue: string;
+  playingPreview: string;
+  showSoundFilters: boolean;
+  onSelect: (path: string) => void;
+  onPreview: (path: string) => void;
+}
+
+const SoundsPanel: React.FunctionComponent<SoundsPanelProps> = ({
+  library,
+  currentValue,
+  playingPreview,
+  showSoundFilters,
+  onSelect,
+  onPreview,
+}) => {
+  const folders = library.getAvailableSounds();
+  const libraryGroupPath = library.getPath();
+
+  const [selectedFolder, setSelectedFolder] = useState<SoundFolder>(
+    library.getAllowedFolderForSoundId(currentValue) || folders[0]
+  );
+  const [mode, setMode] = useState<Mode>('packs');
+  const [filter, setFilter] = useState<Filter>('all');
+
+  const currentFolderRef: React.MutableRefObject<HTMLDivElement | null> =
+    useRef(null);
+  const currentSoundRef: React.MutableRefObject<HTMLDivElement | null> =
+    useRef(null);
+
+  const isDefaultSoundSelected = currentValue === library.getDefaultSound();
+
+  const onModeChange = useCallback((value: Mode) => {
+    setMode(value);
+  }, []);
+
+  const onFilterChange = useCallback((value: Filter) => {
+    setFilter(value);
+  }, []);
+
+  // Initial render.
+  useEffect(() => {
+    // If the user has selected a non-default sound, then jump directly to it.
+    // If the default sound is selected, then the ScrollIntoView component will do
+    // an initial scroll instead.
+    if (!isDefaultSoundSelected) {
+      // This timeout allows the initial scroll-to-current-selection to work
+      // when wrapping the content with FocusLock.
+      setTimeout(() => {
+        currentFolderRef.current?.scrollIntoView();
+        currentSoundRef.current?.scrollIntoView();
+      }, 0);
+    }
+  }, [isDefaultSoundSelected]);
+
+  const currentFolderRefCallback = (ref: HTMLDivElement) => {
+    currentFolderRef.current = ref;
+  };
+
+  const currentSoundRefCallback = (ref: HTMLDivElement) => {
+    currentSoundRef.current = ref;
+  };
+
+  let possibleSoundEntries: SoundEntry[] = [];
+  let rightColumnSoundEntries: SoundEntry[] = [];
+
+  if (mode === 'packs') {
+    folders.sort((a, b) =>
+      a.restricted === b.restricted ? 0 : a.restricted ? -1 : 1
+    );
+    possibleSoundEntries = selectedFolder.sounds.map(sound => ({
+      folder: selectedFolder,
+      sound,
+    }));
+  } else {
+    folders.forEach(folder => {
+      folder.sounds.forEach(sound => {
+        possibleSoundEntries.push({folder, sound});
+      });
+    });
+  }
+
+  if (filter === 'all') {
+    rightColumnSoundEntries = possibleSoundEntries.filter(
+      soundEntry => soundEntry.sound.type !== 'preview'
+    );
+  } else {
+    rightColumnSoundEntries = possibleSoundEntries.filter(
+      soundEntry =>
+        soundEntry.sound.type === filter && soundEntry.sound.type !== 'preview'
+    );
+  }
```

Comment:
Non-blocking, but I'd be inclined to crush this down into a single case and put the conditional inline:

```
rightColumnSoundEntries = possibleSoundEntries.filter(
  soundEntry =>
    soundEntry.sound.type !== 'preview' && (soundEntry.sound.type === filter || filter === 'all')
);
```

Reply:
Nice; done.